pkg/src/src/main/scala/edu/berkeley/cs/amplab/sparkr/RRDD.scala

```diff
@@ -312,6 +314,27 @@ private class StringRRDD[T: ClassTag](
   lazy val asJavaRDD : JavaRDD[String] = JavaRDD.fromRDD(this)
 }
 
+private class BufferedStreamThread(in: InputStream, name: String) extends Thread(name) {
+  val ERR_BUFFER_SIZE = 100
+  val lines = new Array[String](ERR_BUFFER_SIZE)
+  var lineIdx = 0
+  override def run() {
+    for (line <- Source.fromInputStream(in).getLines) {
+      lines(lineIdx) = line
+      lineIdx = (lineIdx + 1) % ERR_BUFFER_SIZE
+      System.err.println(line)
+    }
+  }
+
+  def getLines(): String = {
+    (0 until ERR_BUFFER_SIZE).filter { x =>
```

Comment:
With `takeRight` as suggested above we can avoid this kind of circular tracing. Also, is the filter on null looking for non-filled slots (in which case `takeRight` makes sure we don't need to), or is it just the lines returned can be null themselves (in which case the filter is still useful)?